deso/block.go

```diff
@@ -944,6 +956,161 @@ func (node *Node) GetValidatorPKIDFromSubAccountIdentifier(subAccount *types.Sub
+				glog.Error("getStakeOps: stakeAmountNanos was not a uint64")
+				continue
+			}
+			stakeAmountNanosUint64 := stakeAmountNanos.Uint64()
+
+			operations = append(operations, &types.Operation{
+				OperationIdentifier: &types.OperationIdentifier{
+					Index: int64(numOps),
+				},
+				Account: &types.AccountIdentifier{
+					Address:    lib.Base58CheckEncode(prevValidatorEntry.ValidatorPKID.ToBytes(), false, node.Params),
+					SubAccount: &types.SubAccountIdentifier{Address: ValidatorEntry},
+				},
+				Amount: &types.Amount{
+					Value:    strconv.FormatUint(stakeAmountNanosUint64, 10),
+					Currency: &Currency,
+				},
+				Status: &SuccessStatus,
+				Type:   OutputOpType,
+			})
+			numOps++
+		}
+	}
+	return operations
+}
+
+func (node *Node) getUnstakeOps(txn *lib.MsgDeSoTxn, utxoOps []*lib.UtxoOperation, numOps int) []*types.Operation {
+	if txn.TxnMeta.GetTxnType() != lib.TxnTypeUnstake {
+		return nil
+	}
+
+	var operations []*types.Operation
+	for _, utxoOp := range utxoOps {
+		if utxoOp.Type == lib.OperationTypeUnstake {
+			prevValidatorEntry := utxoOp.PrevValidatorEntry
+			if prevValidatorEntry == nil {
+				glog.Error("getUnstakeOps: prevValidatorEntry was nil")
+				continue
+			}
+			prevStakeEntries := utxoOp.PrevStakeEntries
+			if len(prevStakeEntries) != 1 {
+				glog.Error("getUnstakeOps: prevStakeEntries was not of length 1")
+				continue
+			}
+			prevStakeEntry := prevStakeEntries[0]
+			realTxMeta := txn.TxnMeta.(*lib.UnstakeMetadata)
+			if realTxMeta == nil {
+				glog.Error("getUnstakeOps: realTxMeta was nil")
+				continue
+			}
+			unstakeAmountNanos := realTxMeta.UnstakeAmountNanos
+			if !unstakeAmountNanos.IsUint64() {
+				glog.Error("getUnstakeOps: unstakeAmountNanos was not a uint64")
+				continue
+			}
+			unstakeAmountNanosUint64 := unstakeAmountNanos.Uint64()
+			// First use an "input" from the ValidatorEntry
+			operations = append(operations, &types.Operation{
+				OperationIdentifier: &types.OperationIdentifier{
+					Index: int64(numOps),
+				},
+				Account: &types.AccountIdentifier{
+					Address:    lib.Base58CheckEncode(prevValidatorEntry.ValidatorPKID.ToBytes(), false, node.Params),
+					SubAccount: &types.SubAccountIdentifier{Address: ValidatorEntry},
+				},
+				Amount: &types.Amount{
+					Value:    strconv.FormatUint(unstakeAmountNanosUint64, 10),
+					Currency: &Currency,
+				},
+				Status: &SuccessStatus,
+				Type:   InputOpType,
+			})
+			numOps++
+			operations = append(operations, &types.Operation{
+				OperationIdentifier: &types.OperationIdentifier{
+					Index: int64(numOps),
+				},
+				Account: &types.AccountIdentifier{
+					Address: lib.Base58CheckEncode(prevStakeEntry.StakerPKID.ToBytes(), false, node.Params),
+					SubAccount: node.getLockedStakeEntrySubAccountIdentifierForValidator(
+						prevValidatorEntry.ValidatorPKID, 0), // TODO: we need some way to get the lockedAtEpochNumber. We'll likely need to expose it via utxo operations.
```

Comment:
This looks like more of a FIXME rather than a TODO. If you don't have lockedAtEpochNumber then your unstakes will overwrite each other, meaning people will lose money I think.

Reply:
Yeah this is fixed in core and I just need to update it. I thought I had updated it. I'll address this later today since the requisite core updates are now in.